src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs

```diff
@@ -132,21 +155,12 @@ internal bool HasReferenceOrLoadDirectives
         }
 
         #region Preprocessor Symbols
-        private bool _hasDirectives;
-        private InternalSyntax.DirectiveStack _directives;
-
-        internal void SetDirectiveStack(InternalSyntax.DirectiveStack directives)
-        {
-            _directives = directives;
-            _hasDirectives = true;
-        }
 
         private InternalSyntax.DirectiveStack GetDirectives()
         {
-            if (!_hasDirectives)
+            if (_directives.IsNull)
             {
-                var stack = this.GetRoot().CsGreen.ApplyDirectives(default);
-                SetDirectiveStack(stack);
+                _directives = GetRoot().CsGreen.ApplyDirectives(default);
```

Comment:
>_directives [](http://example.com/codeflow?start=16&length=11)

It doesn't look like this code is thread safe. An assignment to an arbitrary struct is not guaranteed to be atomic

Reply:
Yeah, I noticed that too - that was already an issue before. I think it's actually ok since the struct only contains a single reference field but probably better to improve this.

Reply:
> that was already an issue before.

Since we are switching behavior to use lazy initialization in more cases, the issue is getting more exposure. We probably should add "InterlockedInitialize" helper to the structure, similar to the one we have for ImmutableArray